### ISSUE TITLE
ObjectId: strings vs native

### ISSUE THREAD

**Issue by lemoer** [jue89/flashcards-backend#1]
Why are you using a strings instead of real object-ids?

I first thought that the idGenerator is not necessary because mongo will create the `_id`-field automatically. So I created my first module:

``` javascript
// Fire me up! 

module.exports = { 
        implements: 'model:users',
        inject: []
}

module.exports.factory = function () {
        return {
                schema: {
                        // I removed the _id field here (should it be mandatory?)
                        name: { mandatory: true, type: 'string' },
                        email: { mandatory: true, type: 'string' },
                        admin: { type: 'boolean' }
                },
                index: [ 'email' ]
        };
}
```

Creating users was fine (so I thought everything is okay), but `GET /users/55f2f4d3a9cf0c593f37db02` failed. By debugging I found out, that mongo expects the following `query-selectors`:

``` javascript
{ "_id": "55f2f4d3a9cf0c593f37db02" } // when using strings
{ "_id": ObjectId("55f2f4d3a9cf0c593f37db02") } // when using native ObjectIds
```

**Comment by jue89** [comment]
Yes, that's exactly the reason why I decided to use (temporarily?) strings instead of real ObjectIDs. I don't like that ObjectIDs can't be properly recognised in JSON format. So I decided just to use the available types; as there are: `string`, `number`, `boolean`, `null`, `array` and `object`. (Did I miss one?) Atm every specialised type (like the included types: `uuid-v4`, `sha1-fingerprint`, `color` and `objectid`) is a string with some regex-magic around it.

Do you understand my problem? Maybe you have a better solution for this problem. Btw: MongoDB seems to be fine with strings as primary key. Or do you have discovered real disadvantages?

**Comment by lemoer** [comment]
Well... One thing that crosses my mind is that native object ids are stored as binaries not as UTF (maybe mongo can optimize that better). Another thing is that it's possible to do something like this: `fancy_document._id.getTimestamp()`. Maybe some people could need this..

Personally I don't need one of this features, but I remember that I was often annoyed when frameworks are not using the right mysql datatypes eg. for their dates or ints. I think that they also thought their way was easier to implement.. ;P

**Comment by jue89** [comment]
The timestamp thing is solved by the "created_at" and "modified_at" fields that are added to every document. And yes, storing the string instead of the ObjectID consumes a little bit more memory. I know this is not a perfect solution, but I don't see the need to dedicate much time to solve this.
